### PR TITLE
Fix issue 1120 - Reindexing crashes on big installations

### PR DESCRIPTION
**Pull request by ttournie**
This is a proposition to fix the following issue #1120 

We are currently using this for our own indexing, the goal is to have a more scalable reindex.

### CODE REVIEW

**Comment by michaelbromley** [comment]
Thank you for this contribution! I will set aside some time in the coming days to properly review it.

File: packages/elasticsearch-plugin/src/indexing/indexer.controller.ts [hunk 1]
```diff
@@ -335,42 +335,97 @@ export class ElasticsearchIndexerController implements OnModuleInit, OnModuleDes
...
+                Logger.verbose(`Will reindex ${totalDeletedProductIds} deleted products`, loggerCtx);
+
+                let deletedProductIds = [];
+                let skip = 0;
+                let finishedDeletedItem = 0;
+                do {
+                    const operations: BulkVariantOperation[] = [];
+                    deletedProductIds = await this.connection
+                        .getRepository(Product)
+                        .createQueryBuilder('product')
```
Comment (michaelbromley) [inline packages/elasticsearch-plugin/src/indexing/indexer.controller.ts]: `skip` & `take` are not being used in this query, so this part is not currently being batched.

**Review by michaelbromley — CHANGES_REQUESTED** [review]
See the comment re skip/take.

This PR looks good - the only thing is that the batched updates are being performed on the "live" index (the one that the `variants` alias is pointing to). I think this means that the live `search` query will return partial results if the reindex job is in progress. Have you not run into this scenario in your application?

File: packages/elasticsearch-plugin/src/indexing/indexer.controller.ts (hunk 1 above)
Comment (ttournie) [reply packages/elasticsearch-plugin/src/indexing/indexer.controller.ts]: My bad I have it on my own code, I missed it when creating the PR. I add it

**Comment by ttournie** [comment]
> See the comment re skip/take.
> 
> This PR looks good - the only thing is that the batched updates are being performed on the "live" index (the one that the `variants` alias is pointing to). I think this means that the live `search` query will return partial results if the reindex job is in progress. Have you not run into this scenario in your application?

Well we do not have any problem no.

Has I understand it the reindex index is created by 'cloning' the one that was used (or direclty on it if it fails creating it). So the result are the same has it was before the reindexing at first and then, it gets updated little by little (here by 2500 products) so the first 2500 product will appears before the last 2500 in the search, yes. But I do not see the issue

**Comment by michaelbromley** [comment]
Whether the partial index is an issue I think depends on your needs and also things like the number of channels and languages, which can increase indexing time and leave your storefront longer in a partially-indexed state.

In any case, after you fixed the skip/take part, this looks good to merge and then I will do a few further tests with getting the full index to update transparently.

**Comment by michaelbromley** [comment]
From a recent Slack conversation:

> Maybe one improvement for the full reindex could be to recreate the indice from scratch (not from the current index).
then index the product that needs to be indexed/variant
then switch from the old index to the new.
Right now there is a lot of 'cleaning' operation when reindexing (soft deleted product for exemple) and has the application lives it can have a lot's of old products that should not need to be an operation (edited) 

> I made some changes in this regard and it's much better, I might update my PR

I'll wait on the review to give you time to add these improvements to the PR.

**Comment by michaelbromley** [comment]
I did some memory profiling on this:

27,255 product variants, running `reindex` mutation.

image


blue lines = existing ElasticSearchPlugin
green lines = ElasticSearchPlugin with the changes in this PR
orange line = DefaultSearchPlugin for comparison

**Comment by michaelbromley** [comment]
Rebased and merged in https://github.com/vendure-ecommerce/vendure/commit/bce86f6259daea74e50a29022eab0cb10c134d43

**Pull request closed without merging**